### PR TITLE
mutex: Raise a ThreadError when detecting a fiber deadlock

### PR DESCRIPTION
**Pull request by casperisfine**
[[Bug #19105]](https://bugs.ruby-lang.org/issues/19105)

If no fiber scheduler is registered and the fiber that owns the lock and the one that try to acquire it
both belong to the same thread, we're in a deadlock case.

Ref: https://bugs.ruby-lang.org/issues/17827#note-10

cc @eregon @ioquatix WDYT?

### CODE REVIEW

File: test/fiber/test_mutex.rb
Comment (eregon) [inline test/fiber/test_mutex.rb:238]: Should also join the thread otherwise it can still be alive in the following tests. `kill` doesn't wait for it.

**Review by eregon — APPROVED** [review]
Looks good to me!

File: test/fiber/test_mutex.rb
Comment (eregon) [inline test/fiber/test_mutex.rb:197]: FWIW the `No live threads left. Deadlock?` error is actually an eFatal IIRC.
But I think using ThreadError as you did seems fine too.

There is some other detection too using that message though, maybe it's good to be consistent with it in the exception class?

Comment (eregon) [reply test/fiber/test_mutex.rb:197]: Probably we also need to respect `GET_THREAD()->vm->thread_ignore_deadlock` (which defaults to false) like in `rb_check_deadlock()`, i.e., the lock could be interrupted by a signal and unblock things.

Comment (casperisfine) [reply test/fiber/test_mutex.rb:197]: > FWIW the `No live threads left. Deadlock?` error is actually an eFatal IIRC.

Interesting, the other deadlock related errors in this function are all `rb_eThreadError`. 

> Probably we also need to respect GET_THREAD()->vm->thread_ignore_deadlock

Done.

Comment (eregon) [reply test/fiber/test_mutex.rb:197]: > Interesting, the other deadlock related errors in this function are all `rb_eThreadError`.

Right, ThreadError seems fair then.

**Review by ioquatix — APPROVED** [review]
I would also suggest adding a test using `Fiber.blocked{}`. We don't always go through the scheduler code path even if a scheduler is present, if the fiber is blocking. I believe in this case, it should still error out, but without writing the code & test myself, I am not 100% certain.

**Comment by casperisfine** [comment]
> I would also suggest adding a test using `Fiber.blocked{}`.

I need to read on that one. I'm not familiar with `Fiber.blocked`.

**Comment by ioquatix** [comment]
`Fiber.blocked` disables the fiber scheduler hooks (by disabling the fiber scheduler for the duration of that call).

**Comment by technicalpickles** [comment]
Would it be possible to get a 3.0 and/or 3.1 release with this fix in it? This is blocking our upgrade from 2.7. Happy to help how I can!

**Comment by eregon** [comment]
@technicalpickles The ticket is already marked for backport. But I don't see how this would unblock an upgrade, it will raise instead of deadlock, but that means the code trying to relock the Mutex on the same thread still won't work correctly.

**Comment by technicalpickles** [comment]
@eregon Apologies! I'm not familiar with how backporting and releases happen. Thanks for confirming!

This will help us flush out where deadlock is happening and help confirm when we've fixed it.

**Comment by casperisfine** [comment]
Yeah, backporting is tracked on Redmine, and handle by the release maintainer for each version:

Capture d’écran 2022-11-09 à 06 18 09

A new version will be cut eventually, but probably not anytime soon. I don't know how hard it is with your infra, but I suggest applying the patch yourself when you compile your Ruby.

**Comment by ojab** [comment]
This broke rspec-support, It has implementation of [reentrant mutex](https://github.com/rspec/rspec-support/blob/main/lib/rspec/support/reentrant_mutex.rb#L32-L45) and [corresponding spec](https://github.com/rspec/rspec-support/blob/main/spec/rspec/support/reentrant_mutex_spec.rb#L31-L53).
Right now it fails on `@mutex.lock unless @mutex.owned?` inside the Fiber because now it raises an exception instead of waiting for the lock.

Any hint how to deal with it?

**Comment by casperisfine** [comment]
@ojab could you provide a small reproduction script? I'll happily look at it.

**Comment by casperisfine** [comment]
Also isn't `ReentrantMutex` basically a `Monitor`?

**Comment by eregon** [comment]
> Also isn't `ReentrantMutex` basically a `Monitor`?

It is AFAIK. But RSpec tries to support very very old Ruby versions (1.8).

I think we just need to change the RSpec spec here, so it accepts that outcome (in addition to the existing one).
That specs would be stuck forever except it gets itself unstuck via Thread#raise.
It seems a pretty artificial case which makes sense as a test but not in real code.

Another option would be to use `Thread.ignore_deadlock = true` in that spec, that's designed for the case a deadlock is solved by Thread#raise or a signal, so it seems fair enough.

Self-note: link to my PR fixing ReentrantMutex on Ruby 3.0: https://github.com/rspec/rspec-support/pull/503

**Comment by casperisfine** [comment]
> RSpec tries to support very very old Ruby versions (1.8)

`Monitor` was present in 1.4; https://github.com/ruby/ruby/blame/a36e0c78c90917c4d5cc78f67b3808913795f264/lib/monitor.rb

Based on the RSpec thread the reason seem to be that they don't want to require anything from `stblib` to not impact the tested code.

**Comment by casperisfine** [comment]
@ojab your repro already fails on 3.0.3. What cause it to fail is not this change but https://bugs.ruby-lang.org/issues/17827

**Comment by ojab** [comment]
oh, right, now I got that rspec-support spec is hanging because there is no `f.resume` there. Thanks for explanation and sorry for the noise.

**Comment by pirj** [comment]
Please accept my apologies if I'm saying some complete nonsense. I wonder if it's an irresolvable deadlock?
What if an attempt to lock the mutex owned by a different Fiber of the same Thread would just yield?

I understand the tradeoff of a possibility of that Fiber may get stuck, instead of failing quickly with a `ThreadError`.

**Comment by eregon** [comment]
@pirj The description already mentions that case with the Fiber scheduler, there is no change when there is a Fiber scheduler.
When there is no Fiber scheduler it's an illegal change semantically to change Fiber on Mutex#lock.

**Pull request merged**
